xbmc/addons/AddonManager.cpp

```diff
@@ -297,29 +298,90 @@ bool CAddonMgr::HasAvailableUpdates()
   return !GetAvailableUpdates().empty();
 }
 
+std::vector<std::shared_ptr<IAddon>> CAddonMgr::GetOrphanedDependencies() const
+{
+  std::vector<std::shared_ptr<IAddon>> allAddons;
+  GetAddonsInternal(ADDON_UNKNOWN, allAddons, OnlyEnabled::YES, CheckIncompatible::YES);
+
+  std::vector<std::shared_ptr<IAddon>> orphanedDependencies;
+  for (const auto& addon : allAddons)
+  {
+    if (IsOrphaned(addon, allAddons))
+    {
+      orphanedDependencies.emplace_back(addon);
+    }
+  }
+
+  return orphanedDependencies;
+}
+
+bool CAddonMgr::IsOrphaned(const std::shared_ptr<IAddon>& addon,
+                           const std::vector<std::shared_ptr<IAddon>>& allAddons) const
+{
+  if (CServiceBroker::GetAddonMgr().IsSystemAddon(addon->ID()) ||
+      !CAddonType::IsDependencyType(addon->MainType()))
+    return false;
+
+  for (const auto& other : allAddons)
+  {
+    const auto& deps = other->GetDependencies();
+    auto it = std::find_if(deps.begin(), deps.end(),
+                           [&](const DependencyInfo& dep) { return dep.id == addon->ID(); });
+    if (it != deps.end())
+      return false;
+  }
+  return true;
+}
+
+std::vector<std::string> CAddonMgr::RemoveOrphanedDependencies() const
+{
+  std::vector<std::string> removedItems;
+
+  for (auto toRemove = GetOrphanedDependencies(); toRemove.size() > 0;
+       toRemove = GetOrphanedDependencies())
+  {
+    for (const auto& dep : toRemove)
+    {
+      bool removeData = CDirectory::Exists("special://profile/addon_data/" + dep->ID());
+      CAddonUnInstallJob removeDependency(dep, removeData, false);
+      if (removeDependency.DoWork())
```

Comment:
Oh really? Misusing a class designed for async executing by directly calling its worker method? This at least looks like a hack, even though it works.

Reply:
yup, looks like a hack to me too. but i don't know better atm...

Reply:
hum, still feels like a hack. but we're doing it similarly since Isengard in
```c
bool CAddonInstallJob::Install(const std::string &installFrom, const AddonPtr& repo)
```
https://github.com/xbmc/xbmc/blob/02e7013889e08d608363f2909ebeccdb9ea3b7c9/xbmc/addons/AddonInstaller.cpp#L807-L813

best bet would be changing the direct call to`DoWork()` into `DoModal()` with no progress indicator

Reply:
I'm not sure what the impact of using DoModal() instead is. Not what a correct alternative would be. @ksooo would you suggest something different?

Reply:
sorry guys, i now see this uninstall logic better fits in `CAddonInstaller` singleton, not `CAddonMgr`... but the main question DoWork(), DoModal() or something completely different remains